framework/include/utils/DependencyResolver.h

```diff
@@ -226,25 +264,59 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
   if (_unique_deps.count(k) > 0)
     return;
   _unique_deps.insert(k);
+  auto insert_it = _depends.insert(k);
 
-  if (dependsOn(value, key))
+  for (auto & pr : _colors)
+    pr.second = WHITE;
```

Comment:
I think you'd only need to reset the connected component. But this is plenty of speedup already :)

Reply:
> component

Do you mean components? Adding a new edge could complete a cycle involving an arbitrarily large number of edges. You would have to reset the color of all the nodes in that cycle in order for this algorithm to detect the cycle. Are you suggesting adding a new method that descends the tree, coloring all nodes it finds WHITE, and then run the `hasCycle` method?

Reply:
yeah I mean components. Like starting from this node to every reachable node.

This could be slower than simply traversing the entirety of colors. So I dont know!

Reply:
I think for resetting colors, what you would have to do is descend from the `value` node pre-new-edge-insertion since the insertion can in theory create a cycle and then you would be descending forever. We could try that at some point and check to see whether it's faster than simply iterating over the container which will be guaranteed to be a linear cost directly proportional to the total size of the `_colors` container. Selectively marking should be proportional to the size of the descent chain which will obviously be strictly less or equal to the size of the `_colors` container ... so it could be faster.